src/widgets/barchart/mod.rs

```diff
@@ -45,40 +49,34 @@ pub struct BarChart<'a> {
     label_style: Style,
     /// Style for the widget
     style: Style,
-    /// Slice of (label, value) pair to plot on the chart
-    data: &'a [(&'a str, u64)],
+    /// vector of groups containing bars
+    data: Vec<Group<'a>>,
     /// Value necessary for a bar to reach the maximum height (if no value is specified,
     /// the maximum value in the data is taken as reference)
     max: Option<u64>,
-    /// Values to display on the bar (computed when the data is passed to the widget)
-    values: Vec<String>,
 }
 
 impl<'a> Default for BarChart<'a> {
     fn default() -> BarChart<'a> {
         BarChart {
             block: None,
             max: None,
-            data: &[],
-            values: Vec::new(),
+            data: Vec::new(),
             bar_style: Style::default(),
             bar_width: 1,
             bar_gap: 1,
-            bar_set: symbols::bar::NINE_LEVELS,
             value_style: Style::default(),
             label_style: Style::default(),
+            group_gap: 0,
+            bar_set: symbols::bar::NINE_LEVELS,
             style: Style::default(),
         }
     }
 }
 
 impl<'a> BarChart<'a> {
-    pub fn data(mut self, data: &'a [(&'a str, u64)]) -> BarChart<'a> {
-        self.data = data;
-        self.values = Vec::with_capacity(self.data.len());
-        for &(_, v) in self.data {
-            self.values.push(format!("{v}"));
-        }
+    pub fn data(mut self, data: impl Into<Group<'a>>) -> BarChart<'a> {
```

Comment:
Given that this is the core method for providing data, and we've changed it from accepting the (str, data) tuple array, it would be good to help the user understand their options for providing `Into<Group>` values.